Add v2 RETURNN config serialization

NeoLegends opened the pull request:
This PR adds an adapted version of the v2 serialization written by @albertz. I intend for this PR to house discussion on what features to support in v2 serialization and what not to support.

The adaptations I made mainly concern themselves around being compatible with existing RETURNN configs that use the v1 serialization mechanisms like the `Import`, `PartialImport`, etc.

I also removed `Dim` serialization support to avoid having to import RETURNN from the sisyphus manager (to improve mgr loading speed). I think we should bring back this feature in some way though, because it's very necessary for RF setups.

michelwi commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1165 @@
+    :return: whether the name is a valid Python identifier name (including attrib name)
+    """
+
+    # Very hacky. I'm sure there is some clever regexp, but I don't find it and too lazy...
```
https://docs.python.org/3/library/stdtypes.html#str.isidentifier ?

NeoLegends commented:
I brought back Dim serialization in a way where RETURNN is not imported in the manager.

NeoLegends commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1272 @@
+        known_modules: Collection[str] = (),
+        sis_path_handling: SisPathHandling = None,
+    ):
+        from returnn.tensor import Dim, batch_dim, single_step_dim
```
You can now specify an explicit `returnn_root` that is added to `sys.path` before the serializer is instantiated, so that imports can be directed to the specified path.

albertz replied:
See my other comment on this. I think you can do this in a safe way (checking `obj.__class__.__module__.startswith("returnn.")` on objects). There should not be a `returnn_root` that you need to specify.

albertz commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1284 @@
+) -> SerializedConfig:
+    """serialize config. see module docstring for more info."""
+    if returnn_root is not None:
+        assert tk.running_in_worker(), "returnn_root only supported in sisyphus worker context"
```
Why? While we currently mostly would use it that way, I don't see that this is really a restriction that is necessary.

NeoLegends commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1284 @@
+    inlining: bool = True,
+    known_modules: Collection[str] = (),
+    extra_sys_paths: Sequence[str] = (),
+    returnn_root: Optional[Union[str, Path]] = None,
```
This is for the other way around: if RETURNN is not within your `sys.path`, how do you use this module? The standard AppTek template works this way (see Haotian's comment), RETURNN is not importable from the recipes, and only used from the `ReturnnTrainingJob` via git clone job. Whether this is right or wrong is a discussion that I consider out-of-scope for this PR.

The serializer wants to import some RETURNN types/primitives like Dim so that those can be serialized properly, if they appear within the config. Although in that case you know that RETURNN was available in the recipes beforehand, so maybe I can make the imports lazy/only import them when the config to be serialized contains RETURNN objects.

albertz replied:
To expand why I think this is doubly bad: This just adds another source for potential errors. And it adds complexity. And it is just not needed.

NeoLegends commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1284 @@
+    return tree.map_structure(_instanciate_delayed_obj, o)
+
+
+def _parse_python(code: Any, name=None):
```
This is no longer there

albertz commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1272 @@
+    return name.isidentifier() and not iskeyword(name)
+
+
+class ReturnnConfigWithNewSerialization(ReturnnConfig):
```
You could argue though that @Icemole is right, that this would supersedes the existing class in functionality. A V2 doesn't need to be compatible in functionality to V1.

albertz commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1243 @@
+            # TODO: are these entries needed for correctness?
+            # "batch_dim": batch_dim,
+            # "single_step_dim": single_step_dim,
+            # "Dim": Dim,
```
I think these are not needed. I would however maybe adapt the code for suggesting the variable name to try to keep the original name here.

NeoLegends commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1243 @@
+
+    @staticmethod
+    def _isinstance_returnn_dim(obj: Any) -> bool:
+        return obj.__class__.__name__ == "Dim" == obj.__class__.__module__.startswith("returnn.")
```
Ah, yes. I also noticed the bug/typo by now due to failing tests. I like the suggestion! Applied.

albertz commented on serialization_v2/serialization.py:
```diff
@@ -353,9 +353,11 @@ def _fill_internal_reserved_names_dict_with_known_globals(self, obj: Any):
-            known_globals.update({"batch_dim": batch_dim, "single_step_dim": single_step_dim, "Dim": Dim})
+            known_globals.update(
+                {"batch_dim": batch_dim, "single_step_dim": single_step_dim, "Dim": Dim, "Tensor": Tensor}
+            )
```
That why I said I would adapt the code for the variable name, that you get these cases here. Maybe in general that classes, global funcs and other global objects always keep their original name.

NeoLegends commented:
<details>
<summary>Example config of reasonable complexity serialized with the latest changes:</summary>

```python
#!returnn/rnn.py

import sys

sys.path.insert(0, "/home/mgunz/setups/2025-03-07--ws-asr/recipe/apptek_asr/..")
import sys

sys.path.insert(
    0, "/home/mgunz/setups/2025-03-07--ws-asr/recipe/apptek_asr/../i6_models"
)
get_model = __import__("functools").partial(
    __import__(
        "apptek_asr.lib.pytorch.networks.conformer_aed",
        fromlist=["ConformerTransformerAedCtcModelV1"],
    ).ConformerTransformerAedCtcModelV1,
    **{
        "sampling_rate": 16000,
        "model_dim": 512,
        "out_dim": 10241,
        "num_heads": 8,
        "num_enc_layers": 12,
        "num_dec_layers": 6,
        "aux_loss_layers": (4, 8),
        "bos_idx": 0,
        "eos_idx": 1,
        "specaug_start": (10480, 31440, 52400),
        "use_rf_init": True,
        "share_embedding": True,
        "logits_bias": False,
        "aux_logits_bias": True,
    }
)


import sys

sys.path.insert(0, "/usr/local/bin")
task = "train"
target = "classes"
learning_rate_file = "learning_rates"
use_horovod = True
behavior_version = 21
file_cache_opts = {"cache_directory": "/ssd/$USER/returnn/file_cache"}
torch_dataloader_opts = {"num_workers": 1}
grad_scaler = None
torch_amp = "bfloat16"
torch_distributed = {"param_sync_step": 100, "reduce_type": "param"}
learning_rate_control = "constant"
batch_size = 3600000
max_seqs = 200
max_seq_length = {"classes": 250, "data": 720000}
from apptek_asr.lib.pytorch.training.optimizer import (
    conformer_aed_weight_decay_blacklist_v2,
)

optimizer = {
    "class": "adamw",
    "epsilon": 1e-16,
    "weight_decay": 0.01,
    "weight_decay_custom_include_check": conformer_aed_weight_decay_blacklist_v2,
}
gradient_clip_global_norm = 5.0
extern_data = {
    "data": {"dim": 1},
    "classes": {
        "dim": 10241,
        "sparse": True,
        "vocab": {
            "class": "SentencePieces",
            "model_file": "/nas/models/asr/artefacts/subword_units/sentencepiece/ES/2025-04-spm_10240-telephony-research-task/10240-nmt_nfkc_cf-sep.spm",
            "alpha": 0.7,
            "enable_sampling": True,
        },
    },
}
from returnn.util.file_cache import CachedFile
from returnn.tensor import Dim
from copyreg import __newobj__
from returnn.datasets.postprocessing import Sequential

eval_datasets_value_dataset_map_seq_stream = __newobj__(Sequential)
ExtractAudioFeaturesV1 = __import__(
    "apptek_asr.lib.data_postprocessing", fromlist=["ExtractAudioFeaturesV1"]
).ExtractAudioFeaturesV1(
    data_key="data",
    data_is_in_codec_form=False,
    sample_rate=16000,
    feat_extract_options={
        "features": "raw",
        "peak_normalization": True,
        "preemphasis": None,
        "pre_process": None,
    },
)
eval_datasets_value_dataset_map_seq_stream_state = {"funcs": (ExtractAudioFeaturesV1,)}
eval_datasets_value_dataset_map_seq_stream.__dict__.update(
    eval_datasets_value_dataset_map_seq_stream_state
)
T_dim_1 = Dim(None, name="T")
F_dim = Dim(1, name="F")
T_dim_2 = Dim(None, name="T")
C_dim = Dim(10241, name="C")
eval_datasets_value_1_dataset_map_seq_stream = __newobj__(Sequential)
eval_datasets_value_1_dataset_map_seq_stream_state = {
    "funcs": (ExtractAudioFeaturesV1,)
}
eval_datasets_value_1_dataset_map_seq_stream.__dict__.update(
    eval_datasets_value_1_dataset_map_seq_stream_state
)
eval_datasets = {
    "corpus.ES.f8kHz.es-filtered.Fisher-batch.2.dev.v2": {
        "class": "MultiProcDataset",
        "buffer_size": 800,
        "dataset": {
            "class": "PostprocessingDataset",
            "dataset": {
                "class": "MetaDataset",
                "datasets": {
                    "features": {
                        "class": "HDFDataset",
                        "files": [
                            CachedFile(
                                "/nas/data/speech/ES/8kHz/Fisher/corpus/batch.2.dev.v2/hdf-raw_wav.8kHz.split-1/Fisher-batch.2.dev.v2.hdf.1"
                            )
                        ],
                        "partition_epoch": 1,
                        "seq_ordering": "sorted_reverse",
                    },
                    "targets": {
                        "class": "PostprocessingDataset",
                        "dataset": {
                            "class": "HDFDataset",
                            "files": [
                                CachedFile(
                                    "/home/mgunz/setups/2025-03-07--ws-asr/work/apptek_asr/hdf/create/BlissToTextHdfJob.nMV2kJd36eqY/output/corpus.hdf.001"
                                )
                            ],
                            "partition_epoch": 1,
                        },
                        "map_outputs": {
                            "classes": {
                                "dims": [Dim(None, name="T")],
                                "sparse_dim": Dim(10241, name="classes"),
                                "vocab": {
                                    "class": "SentencePieces",
                                    "model_file": "/nas/models/asr/artefacts/subword_units/sentencepiece/ES/2025-04-spm_10240-telephony-research-task/10240-nmt_nfkc_cf-sep.spm",
                                    "alpha": 0.7,
                                    "enable_sampling": True,
                                },
                                "dtype": "int32",
                            }
                        },
                        "map_seq": __import__(
                            "apptek_asr.lib.data_postprocessing",
                            fromlist=["DecodeTextAndApplyVocabularyV1"],
                        ).DecodeTextAndApplyVocabularyV1(
                            data_key="data",
                            out_data_key="classes",
                            vocab_options={
                                "class": "SentencePieces",
                                "model_file": "/nas/models/asr/artefacts/subword_units/sentencepiece/ES/2025-04-spm_10240-telephony-research-task/10240-nmt_nfkc_cf-sep.spm",
                                "alpha": 0.7,
                                "enable_sampling": True,
                            },
                        ),
                    },
                },
                "data_map": {
                    "data": ("features", "data"),
                    "classes": ("targets", "classes"),
                },
                "seq_order_control_dataset": "features",
            },
            "map_seq_stream": eval_datasets_value_dataset_map_seq_stream,
            "map_outputs": {
                "data": {"dims": [T_dim_1, F_dim], "dtype": "float32"},
                "classes": {"dims": [T_dim_2], "sparse_dim": C_dim, "dtype": "int32"},
            },
        },
        "num_workers": 1,
        "sharding_method": "dedicated",
    },
    "corpus.ES_US.f8kHz.es-filtered.AAA-batch.1.v1": {
        "class": "MultiProcDataset",
        "buffer_size": 800,
        "dataset": {
            "class": "PostprocessingDataset",
            "dataset": {
                "class": "MetaDataset",
                "datasets": {
                    "features": {
                        "class": "HDFDataset",
                        "files": [
                            CachedFile(
                                "/nas/data/speech/ES_US/8kHz/AAA/corpus/batch.1.v1/20250414-hdf-raw_wav.16kHz.split-1/AAA-batch.1.v1.hdf.1"
                            )
                        ],
                        "partition_epoch": 1,
                        "seq_ordering": "sorted_reverse",
                    },
                    "targets": {
                        "class": "PostprocessingDataset",
                        "dataset": {
                            "class": "HDFDataset",
                            "files": [
                                CachedFile(
                                    "/home/mgunz/setups/2025-03-07--ws-asr/work/apptek_asr/hdf/create/BlissToTextHdfJob.SNlQA1HZoJne/output/corpus.hdf.001"
                                )
                            ],
                            "partition_epoch": 1,
                        },
                        "map_outputs": {
                            "classes": {
                                "dims": [Dim(None, name="T")],
                                "sparse_dim": Dim(10241, name="classes"),
                                "vocab": {
                                    "class": "SentencePieces",
                                    "model_file": "/nas/models/asr/artefacts/subword_units/sentencepiece/ES/2025-04-spm_10240-telephony-research-task/10240-nmt_nfkc_cf-sep.spm",
                                    "alpha": 0.7,
                                    "enable_sampling": True,
                                },
                                "dtype": "int32",
                            }
                        },
                        "map_seq": __import__(
                            "apptek_asr.lib.data_postprocessing",
                            fromlist=["DecodeTextAndApplyVocabularyV1"],
                        ).DecodeTextAndApplyVocabularyV1(
                            data_key="data",
                            out_data_key="classes",
                            vocab_options={
                                "class": "SentencePieces",
                                "model_file": "/nas/models/asr/artefacts/subword_units/sentencepiece/ES/2025-04-spm_10240-telephony-research-task/10240-nmt_nfkc_cf-sep.spm",
                                "alpha": 0.7,
                                "enable_sampling": True,
                            },
                        ),
                    },
                },
                "data_map": {
                    "data": ("features", "data"),
                    "classes": ("targets", "classes"),
                },
                "seq_order_control_dataset": "features",
            },
            "map_seq_stream": eval_datasets_value_1_dataset_map_seq_stream,
            "map_outputs": {
                "data": {"dims": [T_dim_1, F_dim], "dtype": "float32"},
                "classes": {"dims": [T_dim_2], "sparse_dim": C_dim, "dtype": "int32"},
            },
        },
        "num_workers": 1,
        "sharding_method": "dedicated",
    },
}
import functools

sys.path.insert(1, "/home/mgunz/setups/2025-03-07--ws-asr/recipe")
from apptek_asr.meta.returnn_experiments.subword_asr_v1 import (
    _get_subep_split_hdf_oggzip_dataset_v2,
)

train = {
    "class": "DistributeFilesDataset",
    "buffer_size": 800,
    "files": [
        (
            "/nas/data/speech/ES/8kHz/Fisher/corpus/batch.1.train.v2/20250414-hdf-raw_wav.16kHz.split-17/Fisher-batch.1.train.v2.hdf.1",
            "/home/mgunz/setups/2025-03-07--ws-asr/work/apptek_asr/hdf/create/BlissToTextHdfJob.C0qNk9mzWQQ6/output/corpus.hdf.001",
        ),
        (
            "/nas/data/speech/ES/8kHz/Fisher/corpus/batch.1.train.v2/20250414-hdf-raw_wav.16kHz.split-17/Fisher-batch.1.train.v2.hdf.2",
            "/home/mgunz/setups/2025-03-07--ws-asr/work/apptek_asr/hdf/create/BlissToTextHdfJob.C0qNk9mzWQQ6/output/corpus.hdf.002",
        ),
        # ...
    ],
    "get_sub_epoch_dataset": functools.partial(
        _get_subep_split_hdf_oggzip_dataset_v2,
        buf_size=200,
        label_decoding_postprocessors=None,
        num_workers=4,
        partition_epoch=1,
        postprocessing_functions=[
            ExtractAudioFeaturesV1,
            __import__(
                "apptek_asr.lib.data_postprocessing", fromlist=["SpeedPerturbationV1"]
            ).SpeedPerturbationV1(
                data_key="data",
                sample_rate=16000,
                factors={0.7: 1, 0.8: 1, 0.9: 1, 1.0: 1, 1.1: 1},
            ),
            __import__(
                "apptek_asr.lib.data_postprocessing", fromlist=["SeqConcatV1"]
            ).SeqConcatV1(
                classes_key="classes",
                classes_sep=3,
                data_key="data",
                max_num_seqs=3,
                max_seq_len=480000,
                num_features_per_class=None,
                num_seqs_dist=("geometric", 0.3),
                seq_tag_sep=":",
            ),
        ],
        postprocessor_map_outputs={
            "data": {"dims": [T_dim_1, F_dim], "dtype": "float32"},
            "classes": {"dims": [T_dim_2], "sparse_dim": C_dim, "dtype": "int32"},
        },
        ordering_postprocessing_function=__import__(
            "returnn.datasets.postprocessing", fromlist=["LaplaceOrdering"]
        ).LaplaceOrdering(num_seqs_per_bin=1000, length_key="data"),
        source_dataset_ordering="random",
        vocab_size=10241,
        vocab_config={
            "class": "SentencePieces",
            "model_file": "/nas/models/asr/artefacts/subword_units/sentencepiece/ES/2025-04-spm_10240-telephony-research-task/10240-nmt_nfkc_cf-sep.spm",
            "alpha": 0.7,
            "enable_sampling": True,
        },
    ),
    "partition_epoch": 6,
    "seq_ordering": "random",
}
device = "gpu"
log = ["./returnn.log"]
log_verbosity = 5
num_epochs = 600
save_interval = 1
cleanup_old_models = {
    "keep": [
        100,
        125,
        150,
        175,
        200,
        225,
        250,
        275,
        300,
        325,
        350,
        375,
        400,
        425,
        450,
        475,
        500,
        525,
        550,
        575,
        590,
        591,
        592,
        593,
        594,
        595,
        596,
        597,
        598,
        599,
        600,
    ]
}
debug_print_layer_output_template = True
log_batch_size = True
multiprocessing = True
backend = "torch"
use_lovely_tensors = True
log_grad_norm = True
use_train_proc_manager = True
torch_log_memory_usage = True
watch_memory = True
model = "/home/mgunz/setups/2025-03-07--ws-asr/work/i6_core/returnn/training/ReturnnTrainingJob.TaZhQFRcTP65/output/models/epoch"


dynamic_learning_rate = __import__("functools").partial(
    __import__(
        "apptek_asr.lib.pytorch.training.learning_rates",
        fromlist=["dyn_lr_piecewise_linear"],
    ).dyn_lr_piecewise_linear,
    **{
        "piecewise_epochs": [0.0, 270.0, 540.0, 600.0],
        "piecewise_values": [1e-05, 0.001, 1e-05, 1e-06],
    }
)

train_step = __import__("functools").partial(
    __import__(
        "apptek_asr.lib.pytorch.training.aed_ctc_train_step",
        fromlist=["aed_train_step"],
    ).aed_train_step,
    **{
        "aed_loss_scale": 1.0,
        "aux_loss_scales": (1.0, 1.0),
        "label_smoothing": 0.1,
        "label_smoothing_start_epoch": 0,
    }
)


# -*- mode: python; tab-width: 4 -*-
```

</details>

NeoLegends commented:
@albertz Look at the example config, and notice how `returnn.datasets.postprocessing.Sequential` is serialized. That currently goes through the pickle route. Do you think/have any idea how to make serialization for these simple, stateless classes prettier in a generic way?

I added some special-casing to serialize `CachedFile` better (because that really blows up the verbosity of your config depending on how you use it), and I can sure add some for `Sequential` as well, but at some point a more generic solution would be better. Can also be done in a follow-up though.

albertz commented on returnn/config.py:
```diff
@@ -36,6 +37,54 @@ def __repr__(self):
         return self.code
 
 
+def unparse_python(code: Any, name: Optional[str] = None, hash_full_python_code: bool = False) -> str:
```
Why do you have that here?

albertz replied:
Ah sorry, I see that we are now in `returnn/config.py`.

But then my question is, why did you move this code? Why did you touch `__parse_python` at all? This is not really relevant for this PR? Or is it? How?

albertz commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1280 @@
+    """
+
+    @staticmethod
+    def from_cfg(old_returnn_cfg: ReturnnConfig):
```
I think this cannot be done in a safe way.

albertz replied:
I thought mostly about using Python epilog / prolog when making this comment. As you want to keep this, and you accept that this is not safe, then also this function here could stay.

But I would say, only if you really use it. Do you?

albertz replied:
Well, I mean, I did not suggest to remove it. I just wanted to understand the use case. But ok, if you think it's better to remove, then let's keep it removed for now.

albertz commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1280 @@
+    from returnn.util.file_cache import CachedFile
+
+
+def serialize_config(
```
I wonder a bit whether we should make this private, i.e.:

```suggestion
def _serialize_config(
```

Because I never actually made use of this directly. It was always via `ReturnnConfigWithNewSerialization`.

(If we make this private, maybe we should also:
* move `ReturnnConfigWithNewSerialization` to the top here
* make `SerializedConfig` private
* adapt the module docstring

)

albertz commented:
There are some other things I noticed from your example. I will make separate comments for each (so that I can mark each as resolved individually).

```python
sys.path.insert(0, "/home/mgunz/setups/2025-03-07--ws-asr/recipe/apptek_asr/..")
```
The path here is not normalized. Should we maybe normalize it to `"/home/mgunz/setups/2025-03-07--ws-asr/recipe"`?

albertz commented:
```python
import sys
...
import sys
```
It seems that `import sys` is done again and again. That seems like a bug. That should not happen.

albertz commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1280 @@
+        self.add_sys_path(mod_path)
+        self.known_modules.add(mod_name)
+
+    def add_sys_path(self, path: str, *, recursive: bool = True):
```
```suggestion
    def add_sys_path(self, path: str, *, recursive: bool = True, code_comment: str = ""):
```
And then below add that `code_comment` as a prefix to the `PyCode` `py_code` string.

albertz replied:
(This is an addition I made in the meantime after you made the PR.)

albertz commented:
I noticed, you have basically 4 times a dim with the name "T". Did you do that intentionally?
```python
T_dim_1 = Dim(None, name="T")
T_dim_2 = Dim(None, name="T")
```
and also two times `Dim(None, name="T")` deep inside some dict.

albertz commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1280 @@
+        if isinstance(value, PartialImport):
+            return self._serialize_partial_import(value, name)
+        if isinstance(value, Import):
+            return self._serialize_import(value, name)
```
Despite the discussion whether you should or should not use those here for `ReturnnConfigWithNewSerialization`, I wonder, why do you need to have this code here at all? Will you even ever get here? Because you have called `instanciate_delayed_copy` already on the config, which should have resolved all of that, or not?

NeoLegends replied:
See https://github.com/rwth-i6/i6_core/pull/601#issuecomment-3540972389

albertz replied:
To answer myself: This is because the `instanciate_delayed_copy` was also modified to be inconsistent to `instanciate_delayed` in that it would *not* handle `SerializerObject`.

albertz commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1272 @@
+
+
+@dataclass
+class SerializedConfig:
```
I think my idea was at some point that I could reuse the existing `ReturnnConfig` and put this whole thing there into the python epilog or so. So then one would not need `ReturnnConfigWithNewSerialization`. But then, I added `ReturnnConfigWithNewSerialization`, and I found this to be much easier and cleaner to use.

So yes, we can maybe remove the whole thing, if you also don't think this is useful / necessary.

`serialize_config` could just return `list(serializer.assignments_dict_by_idx.values())` directly.

albertz commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1280 @@
+        self.assignments_dict_by_idx: Dict[int, PyCode] = {}  # idx -> code
+        self.assignments_dict_by_value_by_type: Dict[type, Dict[Any, PyCode]] = {
+            # TODO: is this entry needed for correctness?
+            # Dim: {},
```
I think this is needed, because the `Dim` serialization is actually simplified, and drops some of the deeper structures, including the `same_as` stuff, and then two dim objects which are equal but not the same instance would not be equal anymore after the serialization.

Maybe you can add such code below:
```python
if "returnn" in sys.modules:
    from returnn.tensor import Dim

    self.assignments_dict_by_value_by_type[Dim] = {}
```

albertz commented:
Btw, another thing that I added, that was somewhat useful:
```python

_in_serialize_config_flag = False


def in_serialize_config() -> bool:
    """
    Whether we are currently inside a serialize_config call.
    """
    return _in_serialize_config_flag


@contextmanager
def _set_in_serialize_config_ctx(value: bool = True):
    global _in_serialize_config_flag
    old_value = _in_serialize_config_flag
    _in_serialize_config_flag = value
    try:
        yield
    finally:
        _in_serialize_config_flag = old_value
```

And in `serialize_config`, I used:
```python
...
with _set_in_serialize_config_ctx():
    serializer.work_queue()
    if inlining:
        serializer.work_inlining()
...
```

See my `i6_experiments.users.zeyer.returnn.delayed_dim.DelayedReduceDim` as an example for how I use this. This `DelayedReduceDim` is an alternative to the `_DelayedDim` that I see sometimes being used, but with one important difference: `_DelayedDim` (and other `DelayedBase` objects) are resolved via `instanciate_delayed_copy`. This happens usually in the `create_files` Sisyphus job task. `DelayedReduceDim` (and other uses of `in_serialize_config`) can implement the `__reduce__` logic in such a way that the actual `Dim` and its opts are only created really when RETURNN is run (in the `run` Sisyphus job task). You get the same effect also when you use `CallImport` and co in the Python prolog/epilog.

It would be good if we can also add this here.

NeoLegends commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1280 @@
+        if isinstance(value, Import):
+            return self._serialize_import(value, name)
+        if isinstance(value, (Checkpoint, PtCheckpoint)):
+            return self._serialize_checkpoint(value)
```
Generally I expect `ReturnnConfig(config={"load", returnn_training_job.out_checkpoints[420]})` to work. And I think before this this caused some serialization issues.

But maybe it was a different bug/incorrectly specified path handling/delayed value instantiation. Let me check if this is still necessary.

albertz commented:
> > The path here is not normalized. Should we maybe normalize it to "/home/mgunz/setups/2025-03-07--ws-asr/recipe"?
> 
> Again, I think this is an AppTek thing. It just injects `{path}/..` and does not normalize that concatenation

Yes, but we could normalize it here. So that was my question. Should we maybe normalize it? (Or you mean because this comes via python epilog, this cannot be done here? I was mostly thinking about the `sys.path` modifications that we do here.)

NeoLegends commented:
Yes this comes from python_prolog/python_epilog.

albertz commented:
> avoiding imports from `torch` in the mgr was to allow patterns like:
> 
> ```
> config = ReturnnConfig({
>   "dynamic_learning_rate": PartialImport(code_object_path="i6_experiments.my.lr.function", hashed_arguments={...})
> })
> ```
> 
> and then the serializer will assign the imported value to the config key that you specified.

Did you actually test this? Because, as said, currently I think it would run `instanciate_delayed_copy` before the serialization. And the `instanciate_delayed_copy` will replace the object just by a string. So the serialization logic will never see the `PartialImport` object? So this will just not work?

albertz commented:
> I generally agree now that the instantiate_delayed should go away, or rather that the functionality should be moved into the serializer. But then again this then introduces another break of the "behaves just like pickle" rule. Those delayed objects kind of break this rule anyway though.

But I argued that this functionality should *not* be part of the serializer. The serializer should just exactly behave like pickle.

NeoLegends commented:
Sorry I edited my comment now. Since they behave like a stand-in for a value that is lazily computed, I wonder what the value is in serializing the "stand-in" part if the value they stand for is "ready". I think we should simply serialize their value like pickle, but not the container (the delayed object) itself.

albertz commented:
> This is an actual example.

No? `postprocessor` is not a valid RETURNN option. RETURNN will ignore this.

NeoLegends commented:
I read RETURNN sources and from them I see that if we have `load = PtCheckpoint(...)` then I expect it to break. E.g. the torch-engine calls `checkpoint.endswith(...)` and this function does not (obviously) exist on the checkpoint class.

Are you sure that serializing the checkpoints without going through their repr will work correctly? `instanciate_delayed` does not handle them I think.

NeoLegends commented:
> which again goes against the pickle principle.

Maybe this is not a problem after all since we _do_ resolve delayed objects before invoking the serializer, so does it really matter if the serializer does this as well?

NeoLegends commented:
> Yes. As said, this is the example I gave before for PtCheckpoint. And this is just one example. You can think of many similar examples.

Ah, yes, right. So that means we actually cannot have both some values be serialized properly and use delayed objects to instanciate them only when needed.

I am wondering if all those use cases only come up because we use these ill-defined wrapper objects. Maybe `(Pt)Checkpoint` should instead be refactored to work with any `PathLike` (which includes `str`) and not just `sisyphus.Path`, or maybe it should never become part of the config?

You say right now "keep the wrapper objects in the config, RETURNN can handle it", but is that a good default choice? Can it really handle these objects in all/most of the cases?  My experience tells me that we generally don't use the wrapper objects as "interface object" to other tools, but always use the raw underlying `str` representation of the paths. So people might expect something else. E.g. when writing RASR-configs we also special-case the wrapper objects and print the path as str 🤔.

And so the "flipped" solution to this problem is to special case _all_ the wrapper objects that we have (`Delayed*`, Path, Checkpoints) and never serialize them in their `i6_core`-/python-form, but always as `str`, and then possibly provide escape hatches the other way if people really want to preserve the pythonic/class representation.

albertz commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1146 @@
+        self.check_consistency()
+
+        config = instanciate_delayed(self.config)
+        post_config = instanciate_delayed(self.post_config)
```
```suggestion
        config = instanciate_delayed_copy(self.config)
        post_config = instanciate_delayed_copy(self.post_config)
```

NeoLegends commented on serialization_v2/serialization.py:
```diff
@@ -212,8 +169,8 @@ def _serialize(self) -> str:
-        config = _instanciate_delayed_copy(self.config)
-        post_config = _instanciate_delayed_copy(self.post_config)
+        config = instanciate_delayed(self.config)
+        post_config = instanciate_delayed(self.post_config)
```
Why would it matter? We're running this in `create_files`, right?

NeoLegends commented:
> I lost track a bit: Do all V1 cases work properly now in V2? Or was there sth which does not work? What exactly does not work?

I think yes, since that was all prolog/epilog + `CodeWrapper`.

albertz commented:
> I just found out the TF checkpoint wrapper class does not have an `__fspath__` method, i.e. it is not `PathLike`. If you simply serialize that and want RETURNN to load it, it will break (currently).

I think this does not really matter for this PR? Then just let's fix the TF checkpoint wrapper.

albertz commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1182 @@
+
+
+@dataclass
+class PyCode(SerializerObject):
```
Maybe this should be private?

```suggestion
class _PyCode(SerializerObject):
```

albertz commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1135 @@
+    py_code: str
+    py_value_repr: Optional[_PyEvalCode] = None
+    is_direct_config_entry: bool = False
+    use_for_hash: bool = False
```
This `use_for_hash` is maybe not needed anymore?

```suggestion
```

NeoLegends commented on serialization_v2/serialization.py:
```diff
@@ -0,0 +1,1182 @@
+
+
+@dataclass
+class PyCode(SerializerObject):
```
I mean yeah, I removed most of it by now. There was a part in the module docstring about the hashing, some handling of whether some part of the PyCode should go into the hash or not and some hash tests. This is now all not there anymore.

NeoLegends commented:
> The path here is not normalized. Should we maybe normalize it to "/home/mgunz/setups/2025-03-07--ws-asr/recipe"?

This path came from an apptek artefact that injects the corresponding code into `python_prolog`, it did not come from the `serialization_v2` code.

I think in the V2 serialization we will never inject unnormalized paths into the config (because we don't do directory traversal there), and I'm hesitant to add logic to modify these manually injected paths. Actually I think it is difficult to do without actually parsing the prolog/epilog.

So I say no: don't do path normalization.

albertz commented on serialization/serialization_v2.py:
```diff
@@ -0,0 +1,1033 @@
+
+See :func:`serialize_config` for a lower level entry point to config dict serialization.
+
+See https://github.com/rwth-i6/i6_experiments/blob/main/users/zeyer/serialization.rst
```
```suggestion
See https://github.com/rwth-i6/i6_core/blob/main/serialization/README.rst
```

albertz reviewed (approved):
See my last small comments. I think otherwise all is fine for merging now.

michelwi commented on returnn/config.py:
```diff
@@ -366,6 +323,93 @@ def _sis_hash(self):
+        # Our automatic mechanism will find direct imports (e.g. i6_experiments).
+        # However, it will not find indirect imports (e.g. sisyphus),
+        # and thus the generated code might fail.
+        # So add all other paths here which we currently have.
```
Are we now sure? -> remove/rewrite comment otherwise, lets try to make sure

michelwi commented on returnn/config.py:
```diff
@@ -366,6 +323,93 @@ def _sis_hash(self):
+        ]
+
+        python_prolog_code = unparse_python(self.python_prolog, hash_full_python_code=True)
+        python_epilog_code = unparse_python(self.python_epilog, hash_full_python_code=True)
```
why is `hash_full_python_code` hardcoded here? It seems the only thing it does is prevent us from using `DelayedBase` objects in prolog/epilog without actually affecting the hash

```suggestion
        python_prolog_code = unparse_python(self.python_prolog, hash_full_python_code=self.hash_full_python_code)
        python_epilog_code = unparse_python(self.python_epilog, hash_full_python_code=self.hash_full_python_code)
```

of course, if the parameter does not make sense in the new implementation, then I would rather add an
```
assert hash_full_python_code == True
```
to the `__init__` of `ReturnnConfigV2`.

But then I am wondering that hashing and serialization should be independent of each other.. so what is the case here?

EDIT: I read below
> There is a :class:`SerializerObject` base class, deriving from Sisyphus :class:`DelayedBase`.

so actually the `# Handle ExternalImports` part does not make sense if we disallow them in the prolog/epilog.

michelwi commented on serialization/serialization_v2.py:
```diff
@@ -0,0 +1,1032 @@
+            rv = self.reduce_cache_by_value_ref[value_ref]
+        else:
+            reduce = getattr(value, "__reduce_ex__", None)
+            reduce_proto = 4  # not sure...
```
> ` # not sure...`

why not? can we make sure?

albertz replied:
Yea, sounds good to change the comment like that.

NeoLegends commented:
🎉
